### ISSUE TITLE
[feature request] Ability to report aggregate metrics in validation

### ISSUE THREAD

**Issue by isaacgerg** [keras-team/keras#10884]
The efficacy of many machine learning problems depend on a metric which can only be evaluated on the entire validation set.  A popular example of such a metric is ROC AUC.  This metric is nonlinear and thus cannot be computed batch-wise.  

A possible concept of operations is to have a callback invoked at the end of validation with passed parameters y_true and y_pred so that such metrics could be evaluated.   The downside of such a function is that y_true and y_pred could be rather large in memory (e.g. y_true is an image).  However, for many problems such as categorical classification, y_true and y_pred will be N < 1e6.

**Comment by briannemsick** [comment]
It exists, see Stateful Metrics.

The metrics test has an example.

**Comment by briannemsick** [comment]
See binary true positives class in

https://github.com/keras-team/keras/blob/master/tests/keras/metrics_test.py

**Comment by brge17** [comment]
(Work github account now), very few metrics actually require the entire validation set and most can be computed in batchwise manner by preserving the state of the previous batches. This is also how TensorFlow computes their metrics. You can look at their AUC implementation to help you write your own AUC stateful metric, or you can wrap their AUC metric if you use TensorFlow backend.

**Comment by isaacgerg** [comment]
Hi Brian,

I dont see anything in the keras.io docs about stateful metrics.  Can you send me the URL?

Thanks,
Isaac

**Comment by brge17** [comment]
There isn't a doc about it,

The basic idea is you inherit from layer, write a call, and reset state function (all of which need the same definition as the example).

Here is another example, it just counts the number of batches (silly I know but look at the progress bar).

```python
import keras
from keras.layers import Input, Dense
from keras.models import Model

import numpy as np

class BatchCounter(keras.layers.Layer):

        def __init__(self, name="batch_counter", **kwargs):
            super(BatchCounter, self).__init__(name=name, **kwargs)
            self.stateful = True
            self.batches = keras.backend.variable(value=0, dtype="int32")

        def reset_states(self):
            keras.backend.set_value(self.batches, 0)

        def __call__(self, y_true, y_pred):
            updates = [
                keras.backend.update_add(
                    self.batches, 
                    keras.backend.variable(value=1, dtype="int32"))]
            self.add_update(updates)
            return self.batches

# Dummy dataset
X = np.ones((50, 1))
y = np.zeros((50, 1))

# Dummy model
inputs = Input(shape=(1,))
outputs = Dense(1)(inputs)
model = Model(inputs=inputs, outputs=outputs)
model.compile(loss="mse", optimizer="adam", metrics=[BatchCounter()])

model.fit(X, y, batch_size=10, epochs=10, validation_data = (X, y))
```

```
50/50 [==============================] - 0s 82us/step - loss: 0.4916 - batch_counter: 5.0000 - val_loss: 0.4836 - val_batch_counter: 5.0000
```

Notice the batch counter is not averaged. It reads 5 (50 samples/10 batch size)

**Comment by isaacgerg** [comment]
Thanks.  I didnt know metrics='' could take a keras.layer.  That didnt seem very intuitive to me until now. Thanks!
